feat(ktableview): add bulk actions feature [KHCP-13106] (#2365)
Kong/kongponents: docs/components/table-view.md

#### Reserved Header Keys

> - `bulkActions` - the column displays checkboxes for each row. Column header displays a checkbox clicking on which will check or uncheck all table rows. Apart from that, KTableView will render a dropdown on the right of the table toolbar directly above the table and you simply need to provide dropdown items via the [`bulk-action-items` slot](#bulk-action-items). Refer to the [`bulk-action-items` slot](#bulk-action-items) for the example.

Reviewer: Unrelated to this line, can we remove "in a nutshell" From the `sortable` interface comment?

> - `bulkActions` - the column displays checkboxes for each row. Column header displays a checkbox clicking on which will check or uncheck all table rows. Apart from that, KTableView will render a dropdown on the right of the table toolbar directly above the table and you simply need to provide dropdown items via the [`bulk-action-items` slot](#bulk-action-items). Refer to the [`bulk-action-items` slot](#bulk-action-items) for the example.

Reviewer: rewording
```suggestion
- `bulkActions` - the column displays individual checkboxes to allow selecting individual rows, while the column header displays a checkbox will check or uncheck all visible table rows. KTableView will render a dropdown on the right of the table toolbar directly above the table and you simply need to provide dropdown items via the [`bulk-action-items` slot](#bulk-action-items). Refer to the [`bulk-action-items` slot](#bulk-action-items) for the example.
```

> - `bulkActions` - the column displays checkboxes for each row. Column header displays a checkbox clicking on which will check or uncheck all table rows and selected rows count. Apart from that, KTableView will render a dropdown on the right of the table toolbar directly above the table and you simply need to provide dropdown items via the [`bulk-action-items` slot](#bulk-action-items). Refer to the [`bulk-action-items` slot](#bulk-action-items) for the example.

Reviewer: Bulk actions applies to the entire table, not to a single column. I'm not sure I understand why this is correct.

Also, what does a user set this value to? Nothing in the description outlines how to use this?

> - `bulkActions` - the column displays checkboxes for each row. Column header displays a checkbox clicking on which will check or uncheck all table rows and selected rows count. Apart from that, KTableView will render a dropdown on the right of the table toolbar directly above the table and you simply need to provide dropdown items via the [`bulk-action-items` slot](#bulk-action-items). Refer to the [`bulk-action-items` slot](#bulk-action-items) for the example.

Author: Bulk actions are enabled for the entire table through providing a header object with a reserved key (`{ ..., key: 'bulkActions' }`). That's how they enable bulk actions column (same goes for `actions` column - if you want to have that column in the table you need to enable it in headers cuz that's how we control which columns are present in the table). The last sentence in this paragraph mentions that example can be found in `bulk-action-items` slot docs section.

- `actions` - the column displays an actions [KDropdown](/components/dropdown) button for each row and displays no label (as if `hideLabel` was `true`; you can set `hideLabel` parameter to `false` to show the label). KTableView will automatically render the actions dropdown and you simply need to provide dropdown items via the [`action-items` slot](#action-items).

:::tip NOTE

> KTableView automatically displays the bulk action checkbox as the first column, and `actions` menu as the last column, when enabled.

Reviewer: ```suggestion
KTableView automatically displays the bulk action checkbox as the first column, and the `actions` menu in the last column, when enabled.
```

</script>
```

### rowBulkAction

> Function for making selected rows unavailable for bulk actions selection conditionally. The function receives row value object as an argument and should return a boolean or an object with two parameters:

Reviewer: ```suggestion
Function that determines if each row enables or disables bulk action selection. The function receives the `row` data object as a parameter and must return a `boolean`, or an object that matches the following interface:
```

```ts
type RowBulkAction = boolean | { disabled: boolean, disabledTooltip?: string }
```

> When function returns a boolean `true`, a row will be enabled. Otherwise, when `false` is returned, it will be disabled. Alternatively, it can return an object (helpful when you need to display a tooltip explaining why the row is unavailable for bulk actions selection).

Reviewer: ```suggestion
When the function returns a boolean value of `true`, bulk action selection will be enabled for the row. 

When `false` is returned, the bulk action checkbox will be disabled. 

If an object is returned, the `enabled` property determines if the row can be selected. The `disabledTooltip` property will show the string as a tooltip message when the disabled checkbox is hovered.
```

<KTableView

>   :row-bulk-action="getRowBulkAction"

Author: This example does have 3 disabled rows...

![Screenshot 2024-09-04 at 4 18 51 PM](https://github.com/user-attachments/assets/20f29aad-980e-4753-9b2d-ac4b05ce0919)

  :data="paginatedData"
  :headers="basicHeaders(false, null, null, true)"
  :pagination-attributes="{ totalCount: basicPaginatedData.length, pageSizes: [5, 10] }"
  @page-change="onPageChange"
  @page-size-change="onPageSizeChange"
  @bulk-actions-select="onBulkActionsSelect"
>
  <template #bulk-action-items>
    <KDropdownItem danger>
      Delete ({{ selectedData.length }} items)
    </KDropdownItem>
  </template>
</KTableView>

```vue
<template>
  <KTableView
    :row-bulk-action="getRowBulkAction"
    :data="paginatedData"
    :headers="headers"
    :pagination-attributes="{ totalCount: tableData.length, pageSizes: [5, 10] }"
  >
    <template #bulk-action-items>
      <KDropdownItem danger>
        Delete ({{ selectedData.length }} items)
      </KDropdownItem>
    </template>
  </KTableView>
</template>

<script setup lang="ts">
import type { RowBulkAction } from '@kong/kongponents'

...

> const getRowBulkAction = (data: Record<string, any>): RowBulkAction => {
>   if (data.id === 2) {
>     return false
>   }
> 
>   if (data.id === 3) {
>     return { disabled: true }
>   }
> 
>   if (data.id === 4) {
>     return {
>       disabled: true,
>       disabledTooltip: 'This row is disabled.',
>     }
>   }
> 
>   return true
> }
> </script>

Reviewer: Since we reference `row` as the parameter, let's use that parameter here in the example
```suggestion
const getRowBulkAction = (row: Record<string, any>): RowBulkAction => {
  if (row.id === 2) {
    return false
  }

  if (row.id === 3) {
    return { disabled: true }
  }

  if (row.id === 4) {
    return {
      disabled: true,
      disabledTooltip: 'This row cannot be selected.',
    }
  }

  return true
}
</script>
```

</script>
```

### rowBulkActionEnabled

> Function for controlling row selection (enabled vs. disabled). Helpful for making some rows unavailable for bulk actions selection. The function receives the row data object as a parameter and must return a `boolean`, or an object that matches the following interface:

Reviewer: ```suggestion
Function for enabling or disabling row selection when `bulkActions` are enabled for the table. Helpful for making some rows unavailable for bulk actions selection. The function receives the row data object as a parameter and must return a `boolean` or an object that matches the following interface:
```

```ts
type RowBulkAction = boolean | { enabled: boolean, disabledTooltip?: string }
```

When the function returns a boolean value of `true`, bulk action selection will be enabled for the row. 

When `false` is returned, the bulk action checkbox will be disabled. 

> If an object is returned, the `enabled` property determines if the row can be selected. The `disabledTooltip` property will show the string as a tooltip message when the disabled checkbox is hovered.

Reviewer: ```suggestion
If an object is returned, the `enabled` property determines if the row can be selected. The `disabledTooltip` value, if provided, will show the string as a tooltip message if the user hovers over the disabled checkbox.
```

</KTableView>
```

### bulk-action-items

> Slot for passing bulk actions dropdown items. See also: [`bulk-actions-select` event](#bulk-actions-select).

Reviewer: ```suggestion
Slot for passing bulk action dropdown items. See also: [`bulk-actions-select` event](#bulk-actions-select).
```

</KTableView>
```

### bulk-action-items

> Slot for passing bulk action dropdown items.

Reviewer: suggestion: can you add a note about this being required if bulk actions are enabled?

> Slot for passing bulk action dropdown items.

Reviewer: question: what happens if the user enables bulkActions but doesn't provide any dropdown actions? Is there a way to check if the slot has content and if not, force bulk actions to be disabled? How should this be handled?

> Slot props:
> - `selectedRows` - array of selected table row objects

Reviewer: nit: formatting
```suggestion
Slot props:

- `selectedRows` - array of selected table row objects
```

</KTableView>

```vue
<template>
  <KTableView
    :data="paginatedData"
    :headers="headers"
    :pagination-attributes="{ totalCount: tableData.length, pageSizes: [5, 10] }"
  >
    <template #bulk-action-items="{ selectedRows }">
      <KDropdownItem danger>
        Delete ({{ selectedRows.length }} items)
      </KDropdownItem>
    </template>
  </KTableView>
</template>

<script setup lang="ts">
import type { TableViewHeader, TableViewData } from '@kong/kongponents'

const headers: Array<TableViewHeader> = [
  {
    key: 'bulkActions', 
    label: 'Bulk actions', 
  },
  ...
]

const tableData = ref<TableViewData>([ ... ])
const paginatedData = tableData.slice(...)
</script>
```

### bulk-actions

Slot for passing custom bulk actions trigger element. Content provided through this slot will replace default bulk actions dropdown.

Slot props:
- `selectedRows` - array of selected table row objects

<KTableView
  :data="basicData"
  :headers="basicHeaders(false, null, null, true)"
  :pagination-attributes="{ totalCount: basicData.length }"
>
  <template #bulk-actions="{ selectedRows }">
    <KButton
      appearance="danger"
      :disabled="!selectedRows.length"
    >

>       Bulk Delete ({{ selectedRows.length }} items selected)

Reviewer: Don't need to say "Bulk"
```suggestion
      Delete {{ selectedRows.length }} {{ selectedRows.length === 1 ? 'item' : 'items' }}
```

Emitted when the user performs sorting, resizes columns or toggles column visibility. Event payload is object of type `TablePreferences` interface (see [`tablePreferences` prop](#tablepreferences) for details).

> ### bulk-actions-select

Author: `select` feels undescriptive to me tbh.. or maybe it's just me

> ### bulk-actions-select

Reviewer: but that's what you're doing? there's not another way to select a row?

> ### bulk-actions-select

Author: How about `bulk-select` or `row-select`? I think in context of a component as big as table is, `select` isn't descriptive enough

> ### bulk-actions-select

Reviewer: row-select is def. better
